src/core/server/plugins/plugins_system.ts

```diff
@@ -0,0 +1,152 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { KibanaCore } from '../../types';
+import { Logger } from '../logging';
+import { Plugin, PluginName } from './plugin';
+import { createPluginInitializerCore, createPluginStartCore } from './plugins_core';
+
+/** @internal */
+export class PluginsSystem {
+  private readonly plugins = new Map<PluginName, Plugin>();
+  private readonly log: Logger;
+  private readonly startedPlugins: PluginName[] = [];
+
+  constructor(private readonly core: KibanaCore) {
+    this.log = core.logger.get('plugins-system');
+  }
+
+  public addPlugin(plugin: Plugin) {
+    this.plugins.set(plugin.name, plugin);
+  }
+
+  public async startPlugins() {
+    const exposedValues = new Map<PluginName, unknown>();
+    if (this.plugins.size === 0) {
+      return exposedValues;
+    }
+
+    const sortedPlugins = this.getTopologicallySortedPluginNames();
```

Comment:
I think topological sorting should happen during discovery instead of as an implementation detail when attempting to kick off the plugin lifecycle. There will be at least one other time when core needs plugins listed in the appropriate topological order: for client-side plugins, we will need to pass a properly sorted array of plugin names to the browser so that the client core can asynchronously load the plugin bundles in effectively the same order that they were loaded on the server, at least as far as dependencies are concerned.

I can't think of a situation where we'd want the results of plugins discovery and not have them already sorted.

What do you think?

Reply:
@epixa have we considered the case when our graph includes plugins without UI (`ui: false` in manifest): so if plugin A (ui: true) depends on plugin B (ui: false) then on the server we should run them in order, but on the client side order doesn't matter it seems and hence graph will look differently.